code/network/lib/Loot/Network/Message.hs

```diff
@@ -68,12 +70,14 @@ handlerDecoded action =
     handler $ \ex (Tagged bs :: Tagged (k,d) ByteString) ->
                 action ex (deserialiseOrFail $ BSL.fromStrict bs)
 
-instance GEq (Sing :: Nat -> *) where
-    geq = testEquality
+instance GEq SNat where
+    geq (SNat :: SNat n) (SNat :: SNat m)
+      | Just Refl <- sameNat (Proxy :: Proxy n) (Proxy :: Proxy m) = Just Refl
+      | otherwise = Nothing
 
-instance D.GCompare (Sing :: Nat -> *) where
+instance GCompare SNat where
     gcompare s1@(SNat :: SNat n) s2@(SNat :: SNat m) =
-        case testEquality s1 s2 of
+        case geq s1 s2 of
           Just Refl -> GEQ
           Nothing ->
             case compare (natVal (Proxy @n)) (natVal (Proxy @m)) of
```

Comment:
Not sure why this whole function checks `testEquality`/`geq` first, but I think it could be reduced to this `case`. 

Moreover it looks like in newer versions of `base` there is a `cmpNat` function that we could have used here and [its implementation](https://hackage.haskell.org/package/base-4.17.0.0/docs/src/GHC.TypeNats.html#cmpNat) is also trusting `compare` and `natVal` to do the right thing :smile:

Reply:
> but I think it could be reduced to this case

It seems that it can't :slightly_frowning_face: 

If we do only `case` here

```haskell
gcompare s1@(SNat :: SNat n) s2@(SNat :: SNat m) =
    case compare (natVal (Proxy @n)) (natVal (Proxy @m)) of
              LT -> GLT
              GT -> GGT
              EQ -> GEQ
                    ^^^
```

There will be the following error

```
Couldn't match type ‘a’ with ‘b’
  Expected: GOrdering a b
    Actual: GOrdering a a
```

AFAIU there is no witness that `n :~: m`, so `GOrdering a a` and `GOrdering a b` are different things. Probably that's why there is an `unsafeCoerce` in `cmpNat` implementation.

Reply:
Ah, I see, I missed that `GEQ` (correctly) expect equality between the two types.

Never mind then, let's leave this as-is